File: 155_Min_Stack.cpp
class MinStack {
private:
    struct element {
        long value;
        long min;
    };

    stack<element> realStack;

public:
    /** initialize your data structure here. */
    // MinStack() {
    //      realStack = new stack<element>;
    // }
    
    void push(int x) {
        element e;
        e.value = x;
        if (!realStack.empty()) {
            if (realStack.top().min >= x) e.min = x;
            else e.min = realStack.top().min;
        }
        else {
            e.min = x;
        }
        realStack.push(e);
    }
    
    void pop() {
        realStack.pop();
    }
    
    int top() {
        return realStack.top().value;
    }
    
    int getMin() {
        return realStack.top().min;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 155_Min_Stack.py
class MinStack:
    class Node:
        def __init__(self, val, minimum, next=None):
            self.val = val
            self.minimum = minimum
            self.next = next
        
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.head = None

    def push(self, x: int) -> None:
        if self.head == None:
            self.head = self.Node(x, x)
        else:
            self.head = self.Node(x, min(self.head.minimum, x), self.head)

    def pop(self) -> None:
        self.head = self.head.next

    def top(self) -> int:
        return self.head.val
        
    def getMin(self) -> int:
        return self.head.minimum


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
